src/angularComponent/ngcOmniboxController.js

```diff
@@ -1,22 +1,64 @@
+import {KEY, isVerticalMovement} from '../coreComponent/keyboard.js';
+
+// Protects against multiple key events firing in a row without disallowing holding down the key
+const KEY_REPEAT_DELAY = 150;
+
 export default class NgcOmniboxController {
   static get $inject() {
-    return [];
+    return ['$document', '$element', '$timeout'];
+  }
+
+  constructor($document, $element, $timeout) {
+    this.$document = $document;
+    this.$element = $element;
+    this.$timeout = $timeout;
+
+    this.highlightedIndex = -1; // -1 means nothing is highlighted
   }
 
-  constructor() {
+  $postLink() {
+    this.inputEl = this.$element[0].querySelector('input[role="combobox"]');
 
+    this._suggestionElements = this.$document[0].getElementsByName(
+      'ngc-omnibox-suggestion-item'
+    ); // Live HTMLCollection so it's always automatically up to date
+
+    // Cached copy as an array we can do transformations on
+    this._cachedNodeList = Array.prototype.slice.apply(this._suggestionElements);
   }
 
   onInputChange() {
-    this.updateSuggestions();
+    this._updateSuggestions();
   }
 
-  updateSuggestions() {
-    this.source({query: this.ngModel}).then((suggestions) => {
-      this.suggestions = suggestions;
-    });
+  onKeyDown($event) {
+    const keyCode = $event.which;
+
+    if (this.hasSuggestions()) {
+
+      if (isVerticalMovement(keyCode)) {
+        $event.preventDefault();
+        $event.stopPropagation();
+      }
+
+      if (!this._keyDownTimeout) {
+        this._handleKeyDown(keyCode);
+      }
+
+      this._keyDownTimeout = this.$timeout(() => this._handleKeyDown(keyCode), KEY_REPEAT_DELAY);
```

Comment:
Is this doing the same thing as `debounce`? If so, I like it a lot because we don't have to add dependencies!! It may help future understanding if we left a comment to the line of:

`// manual debounce so we don't need external dependencies`